ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -382,8 +382,17 @@ public void selectInactiveContainersAndMarkInZK(StatsSnapshot statsSnapshot) {
    * Mark the given {@link Container}s status to INACTIVE in zookeeper.
    * @param inactiveContainerCandidateSet DELETE_IN_PROGRESS {@link Container} set which has been deleted successfully during compaction.
    */
-  private void markContainerInactiveOnZk(Set<Container> inactiveContainerCandidateSet) {
-    // TODO: mark the given containers status to INACTIVE in zookeeper.
+  void markContainerInactiveOnZk(Set<Container> inactiveContainerCandidateSet) {
+    if (inactiveContainerCandidateSet != null) {
+      Map<Short, Account> accountToUpdateMap = new HashMap<>();
+      inactiveContainerCandidateSet.forEach(container -> {
+        container = new ContainerBuilder(container).setStatus(Container.ContainerStatus.INACTIVE).build();
+        short accountId = container.getParentAccountId();
+        Account account = new AccountBuilder(accountToUpdateMap.getOrDefault(accountId, getAccountById(accountId))).addOrUpdateContainer(container).build();
+        accountToUpdateMap.put(accountId, account);
+      });
+      updateAccounts(accountToUpdateMap.values());
```

Comment:
There can be some situations where this update will fail to persist if someone is making other account updates in Nuage concurrently. This is because the update will fail if the account's "snapshotVersion" field does not match. This field is used to make sure that the base account that was modified matches the current account stored in zookeeper so that one update does not erase someone else's change unexpectedly.

I would suggest using a compareAndSet style loop (checking the return value of `updateAccounts` to see if the update went through):
```java
    if (inactiveContainerCandidateSet != null) {
      boolean successful = false;
      while (!successful) {
        Map<Short, Account> accountToUpdateMap = new HashMap<>();
        inactiveContainerCandidateSet.forEach(container -> {
          // start by getting account, and then get container from account to make sure that we are editing the most
          // recent snapshot
          short accountId = container.getParentAccountId();
          Account accountToEdit = accountToUpdateMap.computeIfAbsent(accountId, this::getAccountById);
          Container containerToEdit = accountToEdit.getContainerById(container.getId());
          Container editedContainer =
              new ContainerBuilder(containerToEdit).setStatus(Container.ContainerStatus.INACTIVE).build();
          accountToUpdateMap.put(accountId,
              new AccountBuilder(accountToEdit).addOrUpdateContainer(editedContainer).build());
        });
        // Note that this assumes that unsuccessful updates are always retriable (like snapshot version violations).
        // If there is something unrecoverable wrong, this method will get stuck in this. Max retries may be needed.
        successful = updateAccounts(accountToUpdateMap.values());
      }
    }
```

Reply:
Thanks for the reminding. I've updated to resolve this concurrency issue.